lib/extras/dec_group_jpeg.cc

```diff
@@ -89,47 +100,137 @@ void DequantBlock(const int16_t* JXL_RESTRICT qblock, size_t c,
   }
 }
 
-Status DecodeGroupJpeg(const Image3S& coeffs, size_t group_idx,
-                       const Rect block_rect, const YCbCrChromaSubsampling& cs,
-                       const float* dequant_matrices,
-                       float* JXL_RESTRICT group_dec_cache, size_t thread,
-                       RenderPipelineInput& render_pipeline_input) {
-  HWY_ALIGN float* const block = group_dec_cache;
-  HWY_ALIGN float* const scratch_space = block + kDCTBlockSize;
-
-  size_t hshift[3] = {cs.HShift(0), cs.HShift(1), cs.HShift(2)};
-  size_t vshift[3] = {cs.VShift(0), cs.VShift(1), cs.VShift(2)};
-
-  static constexpr float kDefaultQuantBias[4] = {
-      1.0f - 0.05465007330715401f,
-      1.0f - 0.07005449891748593f,
-      1.0f - 0.049935103337343655f,
-      0.145f,
-  };
-
-  for (size_t c = 0; c < 3; ++c) {
-    ImageF* rpbuffer = render_pipeline_input.GetBuffer(c).first;
-    Rect rect = render_pipeline_input.GetBuffer(c).second;
-    size_t xsize_blocks = DivCeil(block_rect.xsize(), 1 << hshift[c]);
-    size_t ysize_blocks = DivCeil(block_rect.ysize(), 1 << vshift[c]);
-    size_t offset = 0;
-    for (size_t by = 0; by < ysize_blocks; ++by) {
-      float* JXL_RESTRICT idct_row = rect.Row(rpbuffer, by * kBlockDim);
-      size_t idct_stride = rpbuffer->PixelsPerRow();
-      for (size_t bx = 0; bx < xsize_blocks; ++bx) {
-        const int16_t* qblock = &coeffs.PlaneRow(c, group_idx)[offset];
-        offset += kDCTBlockSize;
-        DequantBlock(qblock, c, dequant_matrices, kDefaultQuantBias, block);
-        // IDCT
-        float* JXL_RESTRICT idct_pos = idct_row + bx * kBlockDim;
-        // JPEG XL transposes the DCT, JPEG doesn't.
-        Transpose<8, 8>::Run(DCTFrom(block, 8), DCTTo(scratch_space, 8));
-        TransformToPixels(AcStrategy::DCT, scratch_space, idct_pos, idct_stride,
-                          block);
-      }
+void DecodeJpegBlock(const int16_t* qblock, size_t c,
+                     const float* JXL_RESTRICT dequant_matrices,
+                     float* JXL_RESTRICT scratch_space,
+                     float* JXL_RESTRICT output, size_t output_stride) {
+  HWY_ALIGN float* const block = scratch_space + kDCTBlockSize;
+  DequantBlock(qblock, c, dequant_matrices, kDefaultQuantBias, scratch_space);
+  // JPEG XL transposes the DCT, JPEG doesn't.
+  Transpose<8, 8>::Run(DCTFrom(scratch_space, 8), DCTTo(block, 8));
+  TransformToPixels(AcStrategy::DCT, block, output, output_stride,
```

Comment:
We could also allocate the scratch space on the stack here since IIRC we don't need more than 64 floats or so.

Reply:
Here block and scratch_space are both pre-allocated temp space so we would need 128 floats, but that could still be ok. I will try this out in a different PR.